### ISSUE TITLE
Support for output figure captions in Notedown

### ISSUE THREAD

**Issue by jni** [elegant-scipy/elegant-scipy#239]
We need to support captions, which Notebooks currently don't. Here's my proposal for how to do it:

(1) similar to exercises, wrap cells that output a figure with:

```
<!-- captioned begin -->
(code cell)
<!-- caption begin -->
(caption text)
<!-- caption end -->
<!-- captioned end -->
```

The nesting is optional afaic but I thought it might help to make sure we know what a caption applies to.

(2) Then, the script should attach these captions using the "caption" metadata in the output cell, as detailed in Julius Schulz's [blog post](http://blog.juliusschulz.de/blog/ultimate-ipython-notebook#figures). This may require modifying notedown, or adding a step to notebook conversion.

(3) Finally, the captions should get transferred from the notebooks to the HTML, again, either by plugging nbconvert or by transferring after the fact.

Then, and only then, does [this commit](https://github.com/elegant-scipy/elegant-scipy/commit/c954728b16a3f459aa7c239338c8410b4ee8230b) become useful. =P

**Comment by stefanv** [comment]
How about we just hijack the code snippet and render a caption inside of matplotlib?

I don't think any notebook converter can currently produce figure captions for output?

**Comment by jni** [comment]
have a look at the linked blog post. It talks about nbconvert plugins. I think we could ship our own that "inherits" from the existing html one... Check out the section where he talks about his plugin:
http://blog.juliusschulz.de/blog/ultimate-ipython-notebook#converting-notebook-to-latex-pdf

it's to latex/pdf, but presumably a lot of the code could be co-opted for htmlbook.

**Comment by stefanv** [comment]
I spoke to @mpacer who is working on expanding Jupyter for scientific publication; he mentioned that this blog post relies on a few extensions.  I haven't investigated in detail, but I am hesitant to make our workflow even more complicated than it already is.  For now, perhaps, I can simply wrap the outputs in figure tags?

The syntax would then be something like:

```
<!-- caption begin text="This figure illustrates blah di blah" -->
code
<!-- caption end -->
```

**Comment by jni** [comment]
Sure! I didn't realise you could stick arbitrary text in comment tags. And then you'd parse this caption on the html end, attaching it to whatever image/figure is within the tags? I think that makes a lot of sense.

**Comment by jni** [comment]
btw, can @mpacer share with us where Jupyter plans to take all these things? (captions, figure/table references, citations...)

**Comment by mpacer** [comment]
Hello @stefanv and @jni! So that's still being worked out; happy that you'll be part of the discussion. I'm the one primarily trying to push this kind of thing forward and I was thinking primarily in terms of LaTeX/PDF output, so I'm intrigued by the html possibility.

So right now, the alt text for an image in its own paragraph will be treated as a caption if you were to convert from markdown to html via pandoc[^epub]:

> An image occurring by itself in a paragraph will be rendered as a figure with a caption. (In LaTeX, a figure environment will be used; in HTML, the image will be placed in a div with class figure, together with a caption in a p with class caption.) The image’s alt text will be used as the caption.

But I don't think we treat treat alt texts in that fashion even if they are on their own in our standard html exporter, but it seems like a reasonable option that would bring us closer to consistent functionality. Or at least, there should be a way to do this with a preprocessor or filter. It's probably going to be easier to get caption support in nbconvert before it can be supported in the notebook.

As far as figure and table references: if you explicitly create a `\begin{figure}` or `\begin{tabular}` environment in the notebook, you can use `\label{}`s and there should be a way to get `\autoref` to work inline (but I should note inline LaTeX is really tricky so that probably doesn't work right now). If you have any ideas I'm all ears. But if you don't need this to work in a "notebook" interface, than we're golden. 

Note: figure and equation numbering is really hard to handle in an interactive environment because every time something is rerendered it increments its figure/eq number which is undesirable and confusing to a lot of users. 

Already there is support for citations via the citation preprocessor; this is a little out of date and so I'll need to look into updating it but [this notebook about citations](https://nbviewer.jupyter.org/github/jupyter/nbconvert-examples/blob/master/citations/Tutorial.ipynb) should still be valid. I'm going to look into how to specify the name/path of your bib file using a traitlet so that this feature is more useful (https://github.com/jupyter/nbconvert/issues/568)

Beyond that what exactly are you looking for?

**Comment by jni** [comment]
Hi @mpacer! Yes we are definitely happy to be included. I suspect that anyone who's tried to write a book or paper with Notebook will have ideas to contribute here. Perhaps it's worth setting up a mailing list?

> the alt text for an image in its own paragraph will be treated as a caption

This is great for existing figures, but most of our figures are generated within the notebook. The question is how to get a caption for output cells containing plots! I don't see any clean way to do that other than a "caption" cell metadata field, do you?

But, the major fly in the ointment here is that those captions will get lost if the notebook gets re-run (?), which is unacceptable. So is there a way for the output to "inherit" metadata from the input, perhaps?

At any rate, the important thing is for Notebooks/nbconvert to provide *any* canonical, supported way to add captions to output figures. If there is official support, then we can invest in tooling to do markdown to notebook conversions that include figure captions (not only in our own build tools, but also in packages like notedown, which we use).

> Note: figure and equation numbering is really hard to handle in an interactive environment because every time something is rerendered it increments its figure/eq number which is undesirable and confusing to a lot of users.

Why can't this be done in the same way as LaTeX, with a full pass over the whole notebook, if not on each re-render, then at the press of some UI button or menu item?

> Beyond that what exactly are you looking for?

Longer term, cross-document linking (to link between chapters) would be awesome, but I understand that it is also an awesome task. ;)

**Comment by hdashnow** [comment]
Is this issue solved?

**Comment by jni** [comment]
Yep! @stefanv solved it in #246! =) It's a book-specific hack though. Let's keep the discussion going here! =)

**Comment by jni** [comment]
Incidentally, @pelson had a pretty fantastic suggestion: add a text attribute to matplotlib figures (e.g. `._caption`, and override the display logic of matplotlib plots in Jupyter notebooks. The nice thing is that the caption is then visible in the code, whereas now it is hidden in an html comment.

**Comment by yeus** [comment]
I have the followgin solution:

**Comment by mpacer** [comment]
@yeus's approach is more or less what I was working on in the IPython display mechanism. Except what would be even nicer is to not use top level metadata but mimetype specific metadata so that you can have different captions as is appropriate for different output formats. You would accomplish that by placing the dict of info underneath the mimetype in question.

Today a custom exporter (probably with a custom template) for nbconvert can know how to transform that into an appropriate figure with a caption. 

One of the challenges is that as @fperez pointed out getting this kind of an attribute supported in the notebook's live display which should be supported since it will be needed to ensure that people are able to see the content or their caption inline with the figure in question.

Getting all the details to work will be tricky but if you'll be at SciPy next week, this is one of the things I was hoping to work on & I'm happy to work on it with any of you.

**Comment by mpacer** [comment]
I don't think cell metadata would be overwritten, but doing it in this output specific way is more along the lines of what you were asking for anyway since it's reproducible from source.

Re: labels and eq numbers the problem is how mathjax handles each render as a new figure on the page. We need new renders in order for you to edit the cell. On a single pass it will work without issue which is why it can work in nbconvert. But, in the live notebook, we can't just do a single pass without overwriting the logic within mathjax. 

Cross document linking is possible with something like bookbook: https://github.com/takluyver/bookbook
